### PR TITLE
Improve API and clean up on the way

### PR DESCRIPTION
**Pull request by thraaawn**
Improving the API and turning the cables into capabilities proved to be a greater challenge.

I was unable to tackle this without cleaning up some other stuff on the way which makes this
a really rather large pull request. Sorry about that! I will try to give a good summary of the
changes I've made and the reasoning behind them. This PR includes the previous PR #130.

I'd be happy to discuss all of this via voice chat as well.

Overview of all changes:
- API
  - Added Plugin system [PluginRegistry, AnnotatedInstanceUtil, @StorageNetworkPlugin, IStorageNetworkPlugin]
    Allows other mods to get access to some internal object instances, e.g.
    to get the location of the master for a given connectable. [IStorageNetworkHelpers]
  - Renamed some interfaces, enums and classes to have more meaningful names,
    e.g. IFilterItem is now an IItemStackMatcher, FilterDirection is now EnumStorageDirection etc
  - Turned IConnectable into a capability [IConnectable]
    This allows mods to create new cable type blocks, e.g. could be used by EnderIO conduits.
  - Exposed Import/Export/Storage-Cable functionality as a new capability [IConnectableItemAutoIO, IConnectableLink]
    This allows mods to create new storage type cables, e.g. could be used to get direct access to AE2 storages etc
  - Added a few internal classes to the API [DefaultConnectable, DimPos, EnumStorageDirection, EnumUpgradeType]
    This is necessary to have a fully isolated API package that does not require any of the mods main classes.
  - Exposed some network master functionality [INetworkMaster]: Insert/Extract/Iterate
    Allows mods to interact with the network in general without having to iterate its storages on its own.

- Implications that might have introduced bugs and regressions:
  - Rewrite of all of the master update*() methods since they now need to operate on the (new)
    capabilities.
  - Likewise many of the network messages had to be adapted to the capabilities.
  - And the same for block methods like neighborChanged()

- Cleanup
  - Split Block/Tile/Gui/Container*-Cables into separate classes and moved into packages:
    This allows me to more easily work on this iteratively without going nuts.
    Code wise this removes lots of conditions, since blocks don't have to ask what they are first.
    It also gets rid of variables being used for multiple purposes.
    Also added an intermediate BlockCableWithFacing+TileCableWithFacing class to reduce code duplication.
  - Removed server->client sync of neighboring blocks: [UnlistedPropertyBlockNeighbors]
    The client already knows about these and does not need the server for this. This will make
    transitioning to BlockModels instead of TESRs for the cables a lot easier. It also reduces the
    size of the update package.
  - Replaced StackWrapper with vanilla ItemStacks:
    The size only gets corrupted during nbt store/load since it's using a byte there. This can be
    easily circumvented by additionally storing the actual int size besides the stack in the nbt
    data.
  - Replaced BlockPos with DimPos in all locations that require dimensional awareness:
    This is necessary for mods like compact machines which bridge networks/grids/pipes across dimensions
  - Replaced filter List<StackWrapper>s with individual ItemStackHandlers for convenience:
    This gives us lots of the functionality we need for free without us having to implement basic inventory stuff.
  - Moved Item/Block registration over to the newer @ObjectHolder system
  - Turned Upgrade Types and Gui IDs into Enums for type security
  - Renamed variables to better fit what they actually contain

- Features:
  - Compact Machines compatibility by providing an AbstractNullHandler for it

- Left Todo:
  - Test without FastWorkbench
  - Fix minor known bugs (e.g. teach the master how to extract from multiple inventories with a single request)
  - Find more bugs+regressions and fix them

### CODE REVIEW

**Comment by Lothrazar** [comment]
Havent looked fully into it yet, pushed a tiny commit for two things

-startup crash on existing worlds with null pos in DimPos
- remove hard nonoptional dependency on compactmachines3package , looks like was just a minor leftover

**Comment by Lothrazar** [comment]
Good news so far, loading my massive existing test world network is good so far. at least everything seems to run. I can search and it filters the items, i can add and remove items in the GUI.

Fastcraft looks great no issues i can see so far

i can disconnect and reconnect parts of the network and everything refreshes to hide/show those items, so that main part is all perfect

Import cables work in blacklist mode, but not whitelist mode.
Export cables dont work (they only have blacklist mode)

I cant put anything into the filter slots of the cable GUI's  (the middle two rows) 

Actual result: Nothing happens

Expected result: It takes a "ghost copy" of the item stack (this was where StackWrapper used to be i think), so i the player dont lose any items, and if i take it out i dont get anything.

Then, when it goes to perform an import or export (or a storage action on the purple ones) It checks this Filter list and only allows items to pass through that match it

**Comment by Lothrazar** [comment]
Just for my own future reference for testing this stuff, this is using the released 1.5.X version, with a very tiny test network (gif link below)

https://i.imgur.com/Zc4Y655.gifv

File: src/main/java/mrriegel/storagenetwork/api/capability/IConnectableItemAutoIO.java [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+   *
+   * @param connectablePos The position of your block, including the world
+   * @param master         The network master. Use this to e.g. query amount of items.
+   *
+   * @return Whether or not this IConnectableLink should be processed this tick.
+   */
+  boolean runNow(DimPos connectablePos, INetworkMaster master);
+
+  /**
+   * If this block is used with an ioDirection of OUT and has its getSupportedTransferDirection
```
Comment (Lothrazar) [inline src/main/java/mrriegel/storagenetwork/api/capability/IConnectableItemAutoIO.java]: These javadocs are amazing 10/10

File: src/main/java/mrriegel/storagenetwork/api/data/EnumUpgradeType.java [hunk 2]
```diff
@@ -0,0 +1,18 @@
+package mrriegel.storagenetwork.api.data;
+
+public enum  EnumUpgradeType {
```
Comment (Lothrazar) [inline src/main/java/mrriegel/storagenetwork/api/data/EnumUpgradeType.java]: Love this

File: src/main/java/mrriegel/storagenetwork/block/cable/TileCable.java [hunk 3]
```diff
@@ -1,665 +1,34 @@
...
-
-  private boolean canGoThisDirection(EnumFilterDirection way) {
-    if (isStorageEnabled() && !this.getTransferDirection().match(way)) {
-      return false;
-    }
-    return true;
-  }
-
-  private boolean checkWBFilterList(ItemStack stack) {
-    if (this.isWhitelist()) {
```
Comment (Lothrazar) [inline src/main/java/mrriegel/storagenetwork/block/cable/TileCable.java]: This function is super ugly and gross.... but this is one of the places where the whitelist/blacklist feature of the cables comes into play

Comment (Lothrazar) [reply src/main/java/mrriegel/storagenetwork/block/cable/TileCable.java]: Ok i see ItemStackMatcher  now, looks like that is replacing these features, it just needs to be hooked to GUI i assume?

**Comment by thraaawn** [comment]
I'm currently at work. Will look into the bugs and your questions when I get home.

ItemStackMatcher should replace all the whitelist/filter checking. I might have missed something somewhere or wired it up wrongly. Weirdly enough "What does import/export or whitelist/blacklist mean", even though its basically boolean, is still the most complicated part about all of this :thinking:.

File: src/main/java/mrriegel/storagenetwork/apiimpl/AnnotatedInstanceUtil.java [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+
+import mrriegel.storagenetwork.StorageNetwork;
+import mrriegel.storagenetwork.api.IStorageNetworkPlugin;
+import mrriegel.storagenetwork.api.StorageNetworkPlugin;
+import net.minecraftforge.fml.common.Loader;
+import net.minecraftforge.fml.common.discovery.ASMDataTable;
+
+import java.util.*;
+
+public class AnnotatedInstanceUtil {
```
[the remark this answers is not in the document]
Comment (thraaawn) [reply src/main/java/mrriegel/storagenetwork/apiimpl/AnnotatedInstanceUtil.java]: Oopsie, you are right, this is essentially JEI code. Sorry.
I will add a disclaimer to this and all the other mods of mine that are using some sort of this class.

**Comment by thraaawn** [comment]
> What is AbstractNullHandler ? It seems like removing it didnt break anything, all testing below was with that commit in place

The whole class is simply for additional compatibility with Compact Machines 3. It allows the connections through Compact Machines blocks to be visible on the client. Otherwise they work, but look like they would not connect.
To have a cleaner git history it's probably a good idea to remove this in this PR and create a new one after this is merged with only the CM3 compat change.

> Good news so far, loading my massive existing test world network is good so far.

Excellent! My 1.5->1.6 update-test world only had a smaller network. But this gives me confidence that we have tackled most (if not all) backwards compatibility issues.

> Fastcraft looks great no issues i can see so far

Testing this in a dev environment _without_ fastworkbench is a bit tricky the way gradle is currently set up. I'm not versed enough in gradle to know from the top of my head how to only add the dependency for compilation, but skip it for runtime.

> JEI is MIT and you are welcome to copy, but it’s good to include the license and where it came from.

Done. 

>  work in blacklist mode, but not whitelist mode

I will look into this now.

> I cant put anything into the filter slots of the cable GUI's (the middle two rows)

That's weird. Will look into that as well.

**Comment by thraaawn** [comment]
That should fix both of those issues.

**Comment by thraaawn** [comment]
Any more changes you want me to make? How do we want to proceed with this?

**Comment by Lothrazar** [comment]
Hi , im testing it today and this weekend

**Comment by Lothrazar** [comment]
and also i want to try making an addon mod, like an external block that has the capability

**Pull request merged**